feat: Enable prometheus merge as a configuration in the Kyma Istio CR (#1232)
kyma-project/istio: docs/contributor/adr/0010-support-for-enable-prometheusMerge-configuration.md

> # Support for Enable PrometheusMerge Configuration

Reviewer: ```suggestion
# Support for the Enable PrometheusMerge Configuration
```

## Status
Approved

## Context

> There is a need to support the configuration of enabling `prometheusMerge` to be available in the Istio CR. The background behind this need is explained in [this issue](https://github.com/kyma-project/telemetry-manager/issues/1468) and the impact of this feature being enabled was discussed in an [ADR](https://github.com/kyma-project/telemetry-manager/blob/main/docs/contributor/arch/015-impact-of-istio-prometheus-merge-on-metric-pipelines.md) created in the Telemetry Manager repository.

Reviewer: ```suggestion
There is a need to support the configuration of enabling `prometheusMerge` to be available in the Istio CR. [Telemetry issue 1468](https://github.com/kyma-project/telemetry-manager/issues/1468) explains the background, and [Telemetry ADR 015](https://github.com/kyma-project/telemetry-manager/blob/main/docs/contributor/arch/015-impact-of-istio-prometheus-merge-on-metric-pipelines.md) explains the impact of enabling this feature.
```

## Considerations

> Based on the [original ADR](https://github.com/kyma-project/telemetry-manager/blob/main/docs/contributor/arch/015-impact-of-istio-prometheus-merge-on-metric-pipelines.md), the plan was to originally enable the `prometheusMerge` feature by default. But this has caused problems for users deploying prometheus with a scrape loop sidecar. More details can be found in [this PR](https://github.com/kyma-project/istio/pull/1184).

Reviewer: ```suggestion
Based on the [original ADR](https://github.com/kyma-project/telemetry-manager/blob/main/docs/contributor/arch/015-impact-of-istio-prometheus-merge-on-metric-pipelines.md), the original plan was to enable the `prometheusMerge` feature by default. But this has caused problems for users deploying Prometheus with a scrape loop sidecar. For details, see [PR 1184](https://github.com/kyma-project/istio/pull/1184).
```

## Decision

> The enablement of the `prometheusMerge` feature needs to be a configurable option for the users. By default it will be set to `false`, but users will then need to actively set it to `true` and understand the effects of the feature.

Reviewer: ```suggestion
Enabling the `prometheusMerge` feature must be a configurable option for the users. By default, it will be set to `false`, and users must actively set it to `true` and understand the effects of the feature.
```

> To allow for this configuration, the following API is proposed in the CR under the config option, `prometheusMerge` will have the default value of `false`.

Reviewer: ```suggestion
To support this configuration, the following API is proposed in the Istio CR under the `config` option. `prometheusMerge` will have the default value of `false`.
```

```yaml
apiVersion: operator.kyma-project.io/v1alpha2
kind: Istio
metadata:
    name: default
    namespace: kyma-system
spec:
 config:
   telemetry:
     metrics:
       prometheusMerge: false
```

> The `telemetry` and `metrics` fields are introduced here to account for future plans to introduce more feature from the Istio Telemetry API into the CR.

Reviewer: ```suggestion
The `telemetry` and `metrics` fields are introduced here to account for future plans to introduce more features from the Istio Telemetry API into the CR.
```

> User workloads need to be restarted whenever `prometheusMerge` changes, since the prometheus metrics annotations have to be updated with this configuration change.

Reviewer: ```suggestion
User workloads must be restarted whenever `prometheusMerge` changes, because the Prometheus metrics annotations must be updated with this configuration change.
```

## Consequences

> Istio CustomResourceDefinition will be extended with an additional configuration field of `telemetry.metrics.prometheusMerge` that will allow for configuration of the `prometheusMerge` setting in Istio Mesh Config. The field will be an optional configuration with the default value of `false`.

Reviewer: ```suggestion
Istio CustomResourceDefinition will be extended with an additional configuration field of `telemetry.metrics.prometheusMerge`, which supports configuration of the `prometheusMerge` setting in Istio Mesh Config. The field will be an optional configuration with the default value of `false`.
```

The `telemetry` and `metrics` field will show up in the CR when retrieved from the API server even if not explicitly defined in the manifest.

> ### Sample configuration

Reviewer: ```suggestion
### Sample Configuration
```
Title case for headings

> 1.

Reviewer: Is this a numbered list because the configuration steps must happen in this order? 
Or are they alternatives to each other? In that case, a bullet list would be better.

Regardless, I'd put the explanatory sentence above the code snippet instead of below.

```suggestion
- The following example uses the default value of `false` for the `prometheusMerge` option:
```

> ```

Reviewer: for proper indentation in lists, the code block needs blank spaces at the beginning of each line: 2 spaces if it's an unordered list, 3 spaces for a numbered list.

> apiVersion: istio.operator.kyma-project.io/v1alpha2

Reviewer: ```suggestion
  apiVersion: istio.operator.kyma-project.io/v1alpha2
```

...and so forth

kind: Istio
metadata:
  name: default

```

> This will use the default value of `false` for the `prometheusMerge` option.

Reviewer: Putting it above the code

```suggestion
```

> 2.

Reviewer: ```suggestion
- The following example sets the value of `prometheusMerge` to `true`:
```

> ```

Reviewer: For this code block as well, we need 2 (or 3) blank spaces for proper indentation

apiVersion: istio.operator.kyma-project.io/v1alpha2
kind: Istio
metadata:
  name: default
spec:
  config:
    telemetry:
      metrics:
        prometheusMerge: true
```

> This will set the value of `prometheusMerge` to `true`.

Reviewer: ```suggestion
```

> Restarts to user workloads happen when the `prometheusMerge` field in the `lastAppliedConfiguration` of Kyma Istio CR module differs from the current `prometheusMerge` in the CR. Additionally, we check and only restart Pods with incorrect annotations set.

Reviewer: ```suggestion
Restarts to user workloads happen when the `prometheusMerge` field in the `lastAppliedConfiguration` of Kyma Istio CR module differs from the current `prometheusMerge` in the CR. Additionally, we check and only restart Pods that have incorrect annotations.
```

> When `prometheusMerge` is set to `true`, user workloads will only be restarted when they are missing the `prometheus.io/path:

Reviewer: ```suggestion
When `prometheusMerge` is set to `true`, user workloads are restarted only if they are missing the `prometheus.io/path:/stats/prometheus` and `prometheus.io/port: 15020` annotations.
```
Let's not split a sentence in two lines.

> /stats/prometheus` and `prometheus.io/port: 15020` annotations. Note that the port value may change depending on the default [statusPort](https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#ProxyConfig-status_port) configuration.

Reviewer: ```suggestion
The port value may change depending on the default [statusPort](https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#ProxyConfig-status_port) configuration.
```

> When `prometheusMerge` is set to `false`, user workloads will only be restarted when they have the `prometheus.io/path: /stats/prometheus` and `prometheus.io/port: 15020` annotations.

Reviewer: ```suggestion
If `prometheusMerge` is set to `false`, user workloads are restarted only when they have the `prometheus.io/path: /stats/prometheus` and `prometheus.io/port: 15020` annotations.
```
